pkg/probe/controller.go

```diff
@@ -0,0 +1,172 @@
+// its prober status.
+type Controller interface {
+	io.Closer
+	Start()
+	register(p *Probe, initial error)
+	onChange(p *Probe, newStatus error)
+}
+
+type controllerImpl interface {
+	onClose() error
+	onUpdate(newStatus error)
+}
+
+type controller struct {
+	sync.Mutex
+	statuses map[*Probe]error
+
+	name          string
+	closec        chan struct{}
+	donec         chan struct{}
+	probeInterval time.Duration
+	impl          controllerImpl
+}
+
+func (cb *controller) Start() {
+	if cb.closec != nil {
+		close(cb.closec)
+	}
+
+	cb.closec = make(chan struct{})
+	cb.donec = make(chan struct{})
+	cb.impl.onUpdate(cb.status())
+	go func() {
+		// Considering the consumption of the updating status, invoke
+		// onUpdate every half of the specified interval.
+		t := time.NewTicker(cb.probeInterval / 2)
+	loop:
+		for {
+			select {
+			case <-cb.closec:
+				break loop
+			case <-t.C:
+				cb.impl.onUpdate(cb.status())
+			}
+		}
+		close(cb.donec)
+	}()
+}
+
+func (cb *controller) register(p *Probe, initial error) {
+	cb.Lock()
+	defer cb.Unlock()
+	if _, ok := cb.statuses[p]; ok {
+		log.Debugf("%s is already registered to %s", p, cb.name)
+		return
+	}
+	cb.statuses[p] = initial
+}
+
+func (cb *controller) status() error {
+	cb.Lock()
+	defer cb.Unlock()
+	return cb.statusLocked()
+}
+
+func (cb *controller) statusLocked() (err error) {
+	if len(cb.statuses) == 0 {
+		return fmt.Errorf("%s has no emitters", cb.name)
+	}
+	for p, status := range cb.statuses {
+		if status != nil {
+			err = multierror.Append(err, fmt.Errorf("%s: %v", p, status))
+		}
+	}
+	return err
+}
+
+func (cb *controller) onChange(p *Probe, newStatus error) {
+	cb.Lock()
+	defer cb.Unlock()
+	prev := cb.statusLocked()
+	if stat, ok := cb.statuses[p]; !ok || stat == newStatus {
+		return
+	}
+	cb.statuses[p] = newStatus
+	curr := cb.statusLocked()
+	if prev == curr || (prev != nil && curr != nil) {
+		return
+	}
+	if prev == nil && curr != nil {
+		log.Errorf("%s turns unavailable: %v", cb.name, curr)
+	} else if prev != nil {
+		log.Debugf("%s error status: %v", cb.name, curr)
+	}
+}
+
+func (cb *controller) Close() error {
+	cb.Lock()
+	cb.statuses = map[*Probe]error{}
+	close(cb.closec)
+	cb.Unlock()
+	<-cb.donec
+	return cb.impl.onClose()
+}
+
+type fileController struct {
+	path string
+}
+
+// NewFileController creates a new Controller implementation which creates a file
+// at the specified path only when the registered emitters are all available.
+func NewFileController(opt *Options) Controller {
+	name := filepath.Base(opt.Path)
+	return &controller{
+		statuses:      map[*Probe]error{},
+		name:          name,
+		probeInterval: opt.ProbeInterval,
+		impl:          &fileController{path: opt.Path},
+	}
+}
+
+func (fc *fileController) onClose() error {
+	if _, err := os.Stat(fc.path); os.IsNotExist(err) {
+		return nil
+	}
+	return os.Remove(fc.path)
+}
+
+func (fc *fileController) onUpdate(newStatus error) {
+	if newStatus == nil {
+		f, err := os.Create(fc.path)
```

Comment:
Can you add comment saying "If the file already exits, change the mtime of the file."